neuralmonkey/decoders/transformer.py

```diff
@@ -0,0 +1,347 @@
+    TransformerLayer, position_signal)
+from neuralmonkey.logging import log
+from neuralmonkey.nn.utils import dropout
+from neuralmonkey.vocabulary import (
+    Vocabulary, PAD_TOKEN_INDEX, END_TOKEN_INDEX)
+
+# pylint: disable=invalid-name
+TransformerHistories = extend_namedtuple(
+    "RNNHistories",
+    DecoderHistories,
+    [("decoded_symbols", tf.TensorArray),
+     ("self_attention_histories", List[Tuple]),
+     ("inter_attention_histories", List[Tuple]),
+     ("input_mask", tf.TensorArray)])
+# pylint: enable=invalid-name
+
+
+class TransformerDecoder(AutoregressiveDecoder):
+
+    # pylint: disable=too-many-arguments
+    def __init__(self,
+                 name: str,
+                 encoder: Attendable,
+                 vocabulary: Vocabulary,
+                 data_id: str,
+                 # TODO infer the default for these three from the encoder
+                 ff_hidden_size: int,
+                 n_heads_self: int,
+                 n_heads_enc: int,
+                 depth: int,
+                 max_output_len: int = None,
+                 dropout_keep_prob: float = 1.0,
+                 attention_dropout_keep_prob: float = 1.0,
+                 save_checkpoint: str = None,
+                 load_checkpoint: str = None) -> None:
+        check_argument_types()
+        AutoregressiveDecoder.__init__(
+            self,
+            name=name,
+            vocabulary=vocabulary,
+            data_id=data_id,
+            max_output_len=max_output_len,
+            dropout_keep_prob=dropout_keep_prob,
+            save_checkpoint=save_checkpoint,
+            load_checkpoint=load_checkpoint)
+
+        self.encoder = encoder
+        self.ff_hidden_size = ff_hidden_size
+        self.n_heads_self = n_heads_self
+        self.n_heads_enc = n_heads_enc
+        self.depth = depth
+        self.attention_dropout_keep_prob = attention_dropout_keep_prob
+
+        self.encoder_states = get_attention_states(self.encoder)
+        self.encoder_mask = get_attention_mask(self.encoder)
+        self.dimension = self.encoder_states.get_shape()[2].value
+
+        log("Decoder cost op: {}".format(self.cost))
+        self._variable_scope.reuse_variables()
+        log("Runtime logits: {}".format(self.runtime_logits))
+    # pylint: enable=too-many-arguments
+
+    @property
+    def output_dimension(self) -> int:
+        return self.dimension
+
+    @tensor
+    def embedding_matrix(self) -> tf.Variable:
+        # TODO better initialization
+        return tf.get_variable(
+            "word_embeddings", [len(self.vocabulary), self.dimension],
+            initializer=tf.random_uniform_initializer(-0.5, 0.5))
+
+    def embed_inputs(self, inputs: tf.Tensor) -> tf.Tensor:
+        embedded = tf.nn.embedding_lookup(self.embedding_matrix, inputs)
+        length = tf.shape(inputs)[1]
+        return embedded + position_signal(self.dimension, length)
+
+    @tensor
+    def embedded_train_inputs(self) -> tf.Tensor:
```

Comment:
Proč tohle není v předkovi?

Reply:
protože to nepotřebuje - tohle jsou všechny embedovaný slova v čase. v normálnim dekodéru se embeduje vždycky jen jedno v tom while loopu.